### PR TITLE
Update compliments module

### LINKED ISSUE #3465: Compliments module special days config
I asked this on the forums and was advised to log a bug.

**Platform**: RaspberryPi 5 running raspberryPi OS. Current version of MagicMirror software.

When adding config to enable the "special days" feature of the default compliments module it will seemingly add that days comment to the existing array. In my case that array is considerable so special messages such as "Happy Birthday" or "Merry Christmas" etc. are lost in a sea of compliments.

My expectation of this feature would be that on the configured day it shows only the compliments specified in that area of the config. For example:

`"....-12-25": ["Happy Christmas!"]`

I would expect the only compliment on the 25th December to be "Happy Christmas!". That is not the case though. I am currently adding a line of code in the compliments.js file to empty the array on a special day and only include the configured comments:

`// Add compliments for special days`
`for (let entry in this.config.compliments) {`
`  if (new RegExp(entry).test(date)) {`
`    compliments = []; // Empty the array if it's a special day in config.js`
`    Array.prototype.push.apply(compliments, this.config.compliments[entry]);`
`  }`
`}`

This is not necessarily a "bug" as it may well be the intended behaviour, however I would suggest making this clear in the docs and, if I may be so bold, adding a new config option that allows this behaviour if the user wants it. I am not confident in my coding or GitHub skills to suggest such a change but it seems like it could be quite simple.

### PR DESCRIPTION
**Pull request by WallysWellies**
`Fixes #3465`

Add config option `specialDayUnique` that defaults to `false` and causes special day compliments to be added to the existing compliments array. Setting this option to `true` will only show the compliments that have been configured for that day.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,8 @@ _This release is scheduled to be released on 2024-07-01._
 
 ### Added
 
+- [compliments] Added `specialDayUnique` config option, defaults to `false`. When set to `true` it replaces the existing compliments array with only the compliments that have been configured for that day (#3465)
```
Comment (rejas) [inline CHANGELOG.md:14]: Can be shorter: 

```suggestion
- [compliments] Added `specialDayUnique` config option, defaults to `false`. (#3465)
```

a longer explanation should be added in a seperate PR to the MM docs repo: https://github.com/MagicMirrorOrg/MagicMirror-Documentation

**Review by rejas — CHANGES_REQUESTED** [review]
looks good. sorry for moving the goalpost, but might you also be able to add a test for this to the mm tests?

**Comment by WallysWellies** [comment]
I believe I have done what you ask @rejas. Obviously these tests rely on using the new compliments.js file so I assume tests are done against the develop branch?

**Comment by sdetweil** [comment]
where are the actual tests for this PR?  I don't see any update to the e2e or electron test scripts.

**Comment by sdetweil** [comment]
i added my tests to the electron and e2e tests... but npm run  test:??? unit/e2e/electron all show your test didn't run 
```text
 FAIL   unit  tests/unit/modules/default/compliments/compliments_cron_entry.js
  ● Test suite failed to run

    Your test suite must contain at least one test.
```
but this is in both e2e and electron test suites
```
	describe("Feature cron entry in compliments module", () => {
		beforeAll(async () => {
			await helpers.startApplication("tests/configs/modules/compliments/compliments_cron_entry.js", "06 May 2022 17:03:00 GMT");
			await helpers.getDocument();
		});

		it("shows 'just pub time' compliment on friday, between 16:00 and 19:10, between 00-10 minutes after the hour, only on fridays ", async () => {
				//await helpers.startApplication("tests/configs/modules/compliments/compliments_cron_entry.js", "06 May 2022 17:03:00 GMT");
				await expect(doTest(["just pub time"])).resolves.toBe(true);
		});
	});
```

**Comment by khassel** [comment]
> where are the actual tests for this PR? I don't see any update to the e2e or electron test scripts.

there are no tests in this PR only new config's

**Comment by WallysWellies** [comment]
I'm sorry but I've never used GitHub before and I can't find any documentation to explain how to test configs in the way you seem to expect.

I looked in the tests folder and found example configs so I made mine in the same format and I tested them on my local install. All my commits are based on changes I have made on my live instance because I don't know any other way to do it and I wrote them into the appropriate file using the GitHub web interface (hence the repeated issues with spaces rather than tabs!).

If there is guidance on how to contribute to the project with technical steps please point me to it.

**Comment by khassel** [comment]
I don't expect writing tests if you are not experienced with this ...

> If there is guidance on how to contribute to the project with technical steps please point me to it.

... and we have no guide for writing tests at the moment.

So @rejas can we proceed without tests to have the feature in the next release or other ideas?

**Comment by sdetweil** [comment]
I have a testcase for compliments, I can add his.. 

BUT.. I have a failure.. says 'expects(" ")
```js
	describe("Feature cron entry in compliments module", () => {
		describe("Set date and empty compliments for anytime, morning, evening and afternoon", () => {
			it("shows 'just pub time' compliment on friday, between 16:00 and 19:10, between 00-10 minutes after the hour, only on fridays", async () => {
					await helpers.startApplication("tests/configs/modules/compliments/compliments_cron_entry.js"); //, "06 May 2022 17:03:00 GMT"
					await helpers.getDocument();
					await expect(doTest(["just pub time"])).resolves.toBe(true);
			});
		});
	});
```

from the e2e test 
```

**Comment by WallysWellies** [comment]
I fear I am becoming unqualified to add to this discussion but I have double-checked that my local files match those in the commit and I have retested locally and MM is working as I had expected:

* Using `compliments_specialDayUnique_false.js` in place of my `config.js` file causes the special day message to be added to the existing array so I see it randomly cycling through all 4 messages - 3 from the "anytime" array and one from the "....-..-.." array.
* Using `compliments_specialDayUnique_true.js` causes the special day message to replace the existing array so all I get is the one message from the "....-..-.." array.

I am also using the altered **compliments.js** file from the commit.

Sorry if this is becoming burdensome.

**Comment by sdetweil** [comment]
@WallysWellies 
I am learning too... the code runs, bu the tests fail

npm run test:unit

because the files out there are supposed to be the test script, not the config for the test 
the configs are supposed to be in tests/configs/modules/compliments

and there are three kinds of testcases 

unit,,, - for when a module has sub-parts that return data to higher parts.. weather and calendar are like that 
electron - that starts electron and grabs the module content and compares
and 
e2e  whch runs everything

i just discovered this yesterday, cause I haven't built any testcases before..   made a lot of changes, but all before testing 

so you would move your two config.js files from the unit tree to the configs tree
and then edit the compliments_specs.js test case in the electron/modules/ folder

**Comment by sdetweil** [comment]
SO.. I want to do the test as a false result, BUT 

this in the test process does NOT return false, it rejects
```js
expect(complimentsArray).toContain(await elem.textContent());
```
and the test was 
```js
					await expect(doTest(["Special day message"])).resolves.toBe(false);
```
get 
```text
 ● Compliments module › Feature new parm for date events in compliments module › Set date and compliments for anytime, morning, evening and afternoon and one value for date › may show only date event on friday with new parm 

    expect(received).resolves.toBe()

    Received promise rejected instead of resolved
    Rejected to value: [Error: expect(received).toContain(expected) // indexOf·
    Expected value: "Typical message 1"
    Received array: ["Special day message"]]

      77 | 			it("may show only date event on friday with new parm ", async () => {
      78 | 					await helpers.startApplication("tests/configs/modules/compliments/compliments_specialDayUnique_false.js", "06 May 2022 14:03:00 GMT");
    > 79 | 					await expect(doTest(["Special day message"])).resolves.toBe(false);

```
i don't KNOW what the answer WILL be as its random.. I know it SHOULD NOT be 'Special day message'..
and the message is upside down on expect.. it reports as 'expected' what is in the thing being tested, 
not on what it is tested with (in the parm TO the test (the array))..

how do I fix this?.. I don't know the test language at all. 
all the prior tests are true results tests

**Comment by rejas** [comment]
Will thinking about this, I dont think a test case is doable, since we present the compliments at random. So we cannot really check if the special-day message is shown exclusively or not... So maybe we skip tests for this for now ?

**Comment by khassel** [comment]
I will present a test for this later or tomorrow

**Review by rejas — APPROVED** [review]

**Comment by rejas** [comment]
Thanks @WallysWellies (and @khassel of course) for your work and your patience!

**Comment by WallysWellies** [comment]
I appreciate all your assistance and patience with me! It's been an experience dipping my toes into the world of Git and seeing how it's done. I've enjoyed making my small contribution to the project.

**Comment by sdetweil** [comment]
certainly for the positive/true test, ss only the special day message will
occur.  that the new field.

the negative/false pass is the trouble.

**Comment by sdetweil** [comment]
actually the negative  could have ALL the choices to test against as positive. so any random result will work

**Comment by sdetweil** [comment]
actually that works 
```js
	describe("Feature new parm for date events in compliments module", () => {
		describe("Set date and compliments for anytime, morning, evening and afternoon and one value for date", () => {
			it("may show only date event on friday with new parm ", async () => {
					await helpers.startApplication("tests/configs/modules/compliments/compliments_specialDayUnique_false.js", "06 May 2022 14:03:00 GMT");
					await expect(doTest(["Special day message","Typical message 1","Typical message 2","Typical message 3"])).resolves.toBe(true);
			});
		});
	});
```

**Pull request merged**